source/_posts/2024-01-13-leaky-bucket.markdown

```diff
@@ -0,0 +1,108 @@
+---
+title: "Leaky Bucket Rate Limiting"
+kind: article
+created_at: 2024-01-13 19:40:00 UTC
+author: Shivam Singhal
+layout: post
+---
+Hmm…leaky bucket? The first time I heard these words, it made me chuckle. I thought what kind of name is this for a rate limiting strategy.
+
+But upon exploring it, it’s rather a very simple, elegant, and popular strategy used in a lot of production environments.
+
+But what are we rate limiting? It’s obvious that we want to rate limit the incoming requests. But do we rate limit the whole server(due to hardware constraints) or do we rate limit per user?
+
+The answer, well, depends on your business requirement. How do you differentiate between different users? An easy way out for us is that we can rate limit based on the IP Address of the request and that’s what we are going to do.
```

Comment:
There's a jump betweem "what are we rate limiting" to "it's obvious we're limiting requests". Here's an attempt to explain we're limiting requests a bit.

```suggestion
But what are we rate limiting? The answer, well, depends on your business requirement. A very common use case is to rate limit requests from a user of your service in a given period of time.

How do you differentiate between different users? An easy way out for us is that we can rate limit based on the IP Address of the request and that’s what we are going to do.
```